src/include/distributed/metadata_utility.h

```diff
@@ -301,6 +304,29 @@ typedef struct BackgroundTask
 	} __nullable_storage;
 } BackgroundTask;
 
+/* PG18+ needs a registered snapshot before any CatalogTupleInsert or
+* TOAST-able Datum creation; older PG versions ignore these calls. */
+#if PG_VERSION_NUM >= PG_VERSION_18
+
+/* Push, insert, then pop & unregister so HaveRegisteredOrActiveSnapshot() passes */
+#define CATALOG_INSERT_WITH_SNAPSHOT(rel, tup) \
+	do { \
+		Snapshot __snap = RegisterSnapshot(GetTransactionSnapshot()); \
+		PushActiveSnapshot(__snap); \
+		CatalogTupleInsert((rel), (tup)); \
+		PopActiveSnapshot(); \
+		UnregisterSnapshot(__snap); \
+	} while (0)
+
+#else
+
+/* no-op wrapper on older PGs */
+#define CATALOG_INSERT_WITH_SNAPSHOT(rel, tup) \
+	CatalogTupleInsert((rel), (tup))
+
+#endif
+
+
```

Comment:
Better to move this to pg_version_compat file

Reply:
moved